app/services/articles/feeds/weighted_query_strategy.rb

```diff
@@ -0,0 +1,398 @@
+            ON top_articles.id = articles.id
+          WHERE relevance_score > 0
+          ORDER BY articles.published_at DESC;
+        THE_SQL_STATEMENT
+      end
+
+      def determine_oldest_published_at(user:)
+        [(user&.page_views&.second_to_last&.created_at || 7.days.ago) - 18.hours, 7.days.ago].max
+      end
+
+      # We multiply the relevance score components together.
+      def relevance_score_components_as_sql
+        @relevance_score_components.join(" * ")
+      end
+
+      def offset_if_applicable
+        return "" if @page == 1
+
+        " OFFSET #{(@page - 1) * @number_of_results} "
+      end
+
+      # By default, we use all of the possible scoring methods.
+      def default_scoring_configs
+        SCORING_METHOD_CONFIGURATIONS
+      end
+
+      # This method converts the caller provided :scoring_configs into
+      # an array of SQL clause fragments.
+      #
+      # @param scoring_configs [Hash] the caller provided configurations.
+      #
+      # @see SCORING_METHOD_CONFIGURATIONS
+      # @note Be mindful to guard against SQL injection!
+      def configure!(scoring_configs:)
+        @relevance_score_components = []
+
+        # We looping through the possible scoring method
+        # configurations, we're only accepting those as valid
+        # configurations.
+        SCORING_METHOD_CONFIGURATIONS.each_pair do |valid_method_name, default_config|
+          # Don't attempt to use this factor if we don't have user.
+          next if default_config.fetch(:requires_user) && @user.nil?
+
+          # Ensure that we're only using a scoring configuration that
+          # the caller provided.
+          next unless scoring_configs.key?(valid_method_name)
+
+          scoring_config = scoring_configs.fetch(valid_method_name)
+
+          # If the caller didn't provide a hash for this scoring configuration,
+          # then we'll use the default configuration.
+          scoring_config = default_config unless scoring_config.is_a?(Hash)
+
+          @relevance_score_components << build_score_element_from(
+            # Under NO CIRCUMSTANCES should you trust the caller to
+            # provide a valid :clause.  Don't trust them to send a
+            # valid clause.  That's the path of SQL injection.
+            clause: default_config.fetch(:clause),
+            # We can trust the :cases and :fallback a bit more, as we
+            # later cast them to integers and floats.
+            cases: scoring_config.fetch(:cases),
+            fallback: scoring_config.fetch(:fallback),
+          )
+        end
+      end
+
+      # Responsible for transforming the :clause, :cases, and
+      # :fallback into a SQL fragment that we can use to multiply with
+      # the other SQL fragments.
+      #
+      # @param clause [String]
+      # @param cases [Array<Array<#to_i, #to_f>>]
+      # @param fallback [#to_f]
+      def build_score_element_from(clause:, cases:, fallback:)
+        text = "(CASE #{clause}"
```

Comment:
While I know that we control all the values here, I have to say that so much SQL being created by hand with interpolation makes me nervous.

Reply:
I absolutely agree…and am completely open for suggestions.

Reply:
I'd probably start by looking into using some of the methods in [https://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html](https://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html) (it's a module, so you should be able to `include`/`extend` it).